File: backend/routers/simulations.py
"""API routes for simulation sessions."""

from database import get_db
from fastapi import APIRouter, Depends, HTTPException
from schemas.simulation import (
    AutoRespondResponse,
    MessageResponse,
    MessageSend,
    SimulationCreate,
    SimulationResponse,
    SimulationSummary,
)
from schemas.tool_call import ToolCallResponse
from services import simulation_service
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(prefix="/api/simulations", tags=["simulations"])


@router.get("", response_model=list[SimulationSummary])
async def list_simulations(
    db: AsyncSession = Depends(get_db),
) -> list[SimulationSummary]:
    """List all simulation sessions (newest first)."""
    sessions = await simulation_service.list_sessions(db)
    return [SimulationSummary.model_validate(s) for s in sessions]


@router.post("", response_model=SimulationResponse, status_code=201)
async def create_simulation(
    data: SimulationCreate, db: AsyncSession = Depends(get_db)
) -> SimulationResponse:
    """Create a new simulation session."""
    session = await simulation_service.create_session(db, data)
    return SimulationResponse.model_validate(session)


@router.get("/{simulation_id}", response_model=SimulationResponse)
async def get_simulation(
    simulation_id: str, db: AsyncSession = Depends(get_db)
) -> SimulationResponse:
    """Fetch a simulation session with full transcript."""
    session = await simulation_service.get_session(db, simulation_id)
    if not session:
        raise HTTPException(status_code=404, detail="Simulation not found")
    return SimulationResponse.model_validate(session)


@router.get(
    "/{simulation_id}/tool_calls",
    response_model=list[ToolCallResponse],
)
async def get_tool_calls(
    simulation_id: str, db: AsyncSession = Depends(get_db)
) -> list[ToolCallResponse]:
    """Fetch all tool calls for a simulation session."""
    session = await simulation_service.get_session(db, simulation_id)
    if not session:
        raise HTTPException(status_code=404, detail="Simulation not found")
    calls = await simulation_service.get_tool_calls(db, simulation_id)
    return [ToolCallResponse.model_validate(c) for c in calls]


@router.post(
    "/{simulation_id}/messages",
    response_model=MessageResponse,
)
async def send_message(
    simulation_id: str,
    data: MessageSend,
    db: AsyncSession = Depends(get_db),
) -> MessageResponse:
    """Send a user message and get the agent's response."""
    session = await simulation_service.get_session(db, simulation_id)
    if not session:
        raise HTTPException(status_code=404, detail="Simulation not found")

    response = await simulation_service.process_message(db, session, data.content)

    return MessageResponse(
        agent_message=response.agent_message,
        tool_calls=[tc for tc in response.tool_calls],
        escalation=response.escalation,
        stop_reason=response.stop_reason,
    )


@router.post(
    "/{simulation_id}/auto_responses",
    response_model=AutoRespondResponse,
)
async def auto_respond(
    simulation_id: str,
    db: AsyncSession = Depends(get_db),
) -> AutoRespondResponse:
    """Patient simulator generates a message, then the agent responds."""
    session = await simulation_service.get_session(db, simulation_id)
    if not session:
        raise HTTPException(status_code=404, detail="Simulation not found")

    if not session.scenario_name:
        raise HTTPException(
            status_code=400,
            detail="Auto-respond requires a session with a scenario.",
        )

    patient_message, response = await simulation_service.auto_respond(db, session)

    return AutoRespondResponse(
        patient_message=patient_message,
        agent_message=response.agent_message,
        tool_calls=[tc for tc in response.tool_calls],
        escalation=response.escalation,
        stop_reason=response.stop_reason,
    )

File: frontend/src/api/simulations.ts
import client from "./client";

export interface ToolCallOut {
  tool_name: string;
  tool_input: Record<string, unknown>;
  status: string;
  output: Record<string, unknown>;
}

export interface EscalationOut {
  type: string;
  keyword: string;
  action: string;
}

export interface MessageResponse {
  agent_message: string;
  tool_calls: ToolCallOut[];
  escalation: EscalationOut | null;
  stop_reason: string;
}

export interface SimulationSession {
  id: string;
  practice_id: string;
  config_id: string;
  scenario_name: string | null;
  channel_mode: string;
  messages: ChatMessage[];
  outcome: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface SimulationSummary {
  id: string;
  practice_id: string;
  config_id: string;
  scenario_name: string | null;
  channel_mode: string;
  outcome: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface SimulationCreate {
  practice_id: string;
  config_id: string;
  scenario_name?: string | null;
  channel_mode?: string;
}

export interface MessageSend {
  content: string;
}

export async function listSimulations(): Promise<SimulationSummary[]> {
  const res = await client.get<SimulationSummary[]>("/simulations");
  return res.data;
}

export async function createSimulation(
  data: SimulationCreate,
): Promise<SimulationSession> {
  const res = await client.post<SimulationSession>("/simulations", data);
  return res.data;
}

export async function getSimulation(id: string): Promise<SimulationSession> {
  const res = await client.get<SimulationSession>(`/simulations/${id}`);
  return res.data;
}

export async function sendMessage(
  simulationId: string,
  data: MessageSend,
): Promise<MessageResponse> {
  const res = await client.post<MessageResponse>(
    `/simulations/${simulationId}/messages`,
    data,
  );
  return res.data;
}
